earth2studio/data/ecmwf.py

```diff
@@ -0,0 +1,752 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024-2025 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-FileCopyrightText: All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+import functools
+import hashlib
+import os
+import pathlib
+import shutil
+from abc import ABC, abstractmethod
+from collections.abc import Callable
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import Any, Literal
+
+import numpy as np
+import xarray as xr
+from loguru import logger
+from tqdm.asyncio import tqdm
+
+from earth2studio.data.utils import datasource_cache_root, prep_forecast_inputs
+from earth2studio.lexicon import AIFSLexicon, IFSLexicon
+from earth2studio.lexicon.ecmwf import ECMWFOpenDataLexicon
+from earth2studio.utils.imports import (
+    OptionalDependencyFailure,
+    check_optional_dependencies,
+)
+from earth2studio.utils.type import LeadTimeArray, TimeArray, VariableArray
+
+try:
+    import ecmwf.opendata as opendata
+except ImportError:
+    OptionalDependencyFailure("data")
+    opendata = None
+
+logger.remove()
+logger.add(lambda msg: tqdm.write(msg, end=""), colorize=True)
+
+
+@dataclass
+class ECMWFOpenDataAsyncTask:
+    """Small helper struct for Async tasks"""
+
+    data_array_indices: tuple[int, int, int]
+    time: datetime
+    lead_time: timedelta
+    variable: str
+    levtype: str
+    level: str | list[str]
+    modifier: Callable
+
+
+@check_optional_dependencies()
+class ECMWFOpenDataSource(ABC):
```

Comment:
We need an available method for data sources. If you look at other data sources you can find references for this.

Reply:
Oh hmm, I did not realize this. I was thinking that all the data sources had the available method. I suppose its ok to not have in that case.